clarify unexpected unmocked error message

philomates opened the pull request:
for something unmocked like
```clojure
(ns mockfn.scratch
  (:require [mockfn.macros :refer [providing unmocked]]))

(defn bar [x y] 1)
(defn foo []
  (bar "b" 1))

(providing [(bar "a" 1) nil]
 (foo))
```
we used to get the error message

```
Unexpected call to mockfn.scratch$bar@c695421 with args ["b" 1]
```

and this PR changes it to be
```
Syntax error (ExceptionInfo) compiling at (mockfn/macros_test.cljc:125:1).
mockfn.fixtures/one-fn was specified as mocked but none of the cases matched the 1 provided argument(s):
  ["unexpected"]

Either
 - specify a mock case for those arguments
 - specify falling through to the original functionality via a terminal match clause:
  `(mockfn.fixtures/one-fn mockfn.matchers/any-args?) mockfn.macros/fall-through
```

I'm proposing this change because it feels a bit more in-line with the fix which is to add an `unmocked` entry:
```clojure
(providing [(bar "a" 1) nil
            (bar "b" 1) unmocked]
 (foo))
```

additionally:

 - implemented any-args? (appropriate name?) to match irregardless of # of args provided
 - deprecated `unmocked` and replaced it with `fall-through`

philomates committed:
handle plurals in unexpected args

gabrielgiussi commented:
`unexpected unmocked` is not somehow redundant?
How an `unexpected mocked` call would look like?

dchelimsky commented:
~~The function is the unmocked thing, not the call, so how about:~~

~~"An unexpected call to unmocked mockfn.scratch$bar@332f6d6a was made with 2 argument(s): ["b" 1]"~~

dchelimsky commented:
Actually this PR makes me think `unmocked` is a confusing name. It forwards to `calling-original`. How about we rename it to something that indicates it's the original implementation (leaving `unmocked` in place, but deprecated)?

dchelimsky commented:
> Actually this PR makes me think `unmocked` is a confusing name. It forwards to `calling-original`. How about we rename it to something that indicates it's the original implementation (leaving `unmocked` in place, but deprecated)?

To elaborate: the function we're calling `unmocked` is probably mocked with different arguments, so calling it `unmocked` is confusing. I'd like to think about this naming a bit more before changing error messages.

philomates commented:
To be honest, having not used mockfn in a long time I sort of expected that when you mock a function and none of the arg preds are satisfied it would fall-through to the underlying original implementation. Once I get past that and realize I need to tell it to handle the `:else` case explicitly, I don't feel a big difference between names like `unmocked`, `unstubbed`, `calling-original` `original`, etc.

philomates commented:
> `unexpected unmocked` is not somehow redundant?
> How an `unexpected mocked` call would look like?

Totally. This I guess mirrors the fact that I didn't expect to need to provide an `unmocked` fall-through case :)

dchelimsky commented:
> > `unexpected unmocked` is not somehow redundant?
> > How an `unexpected mocked` call would look like?
> 
> Totally. This I guess mirrors the fact that I didn't expect to need to provide an `unmocked` fall-through case :)

`fall-through` is a great option to replace `unmocked`. Or `pass-through`.

philomates commented:
`fall-through` sounds good
How about I update the error to say:
```
An unmocked call to mockfn.scratch$bar@332f6d6a was made with 2 argument(s): ["b" 1]
If you would like this case have the original functionality mark it with `fall-through`
```
add `fall-through` and deprecate `unmocked`?

dchelimsky commented:
> `fall-through` sounds good
> How about I update the error to say:
> 
> ```
> An unmocked call to mockfn.scratch$bar@332f6d6a was made with 2 argument(s): ["b" 1]
> If you would like this case have the original functionality mark it with `fall-through`
> ```
> 
> add `fall-through` and deprecate `unmocked`?

Almost there. I'm not sure about the first sentence because the definition of "unmocked" is unclear. I think most people would see that, look at the fact that they mocked the call to `bar` and scratch their heads over the fact that mocked means "with these specific args only".  I'd like find a terse way of conveying that in the error message. The rest is great, though.

dchelimsky commented:
I also wonder if we could clean up the name of the fn e.g.

```clojure
#'mockfn.scratch/bar was called with unexpected args ["b" 1].
If you would like this case to have the original functionality mark it with `fall-through`
```

Related: I wonder if `fall-through` should have an option to indicate "fall through for all unspecified args". WDYT about that @philomates ?

philomates commented:
Okay, how about?:
```
mockfn.scratch$bar@332f6d6a was specified as mocked but none of the cases matched the 2 provided argument(s): ["b" 1]
Either specify a mock case for those arguments or specify falling through to the original functionality via a terminal match clause:
`(mockfn.scratch$bar@332f6d6a any-args?) fall-through`
```
where we could implement `any-args?` to be a match that doesn't care about what args are or how many there are

dchelimsky commented:
> ```
> mockfn.scratch$bar@332f6d6a was specified as mocked but none of the cases matched the 2 provided argument(s): ["b" 1]
> Either specify a mock case for those arguments or specify falling through to the original functionality via a terminal match clause:
> `(mockfn.scratch$bar@332f6d6a any-args?) fall-through`
> ```

It's like a whole tutorial in a docstring. <3 <3 <3 <3 <3 <3 <3 <3

philomates committed:
clarify arg name & use helper to concretize some higher-order fns

philomates committed:
implement any-args? matcher

philomates committed:
Deprecate `unmocked` and replace with `fall-through`

philomates committed:
update missing case message

philomates committed:
remove Unbound artifact from error messages

philomates commented:
cool the message is a little verbose but I think worth it. It looks like:
```
Syntax error (ExceptionInfo) compiling at (mockfn/macros_test.cljc:125:1).
mockfn.fixtures/one-fn was specified as mocked but none of the cases matched the 1 provided argument(s):
  ["unexpected"]

Either
 - specify a mock case for those arguments
 - specify falling through to the original functionality via a terminal match clause:
  `(mockfn.fixtures/one-fn mockfn.matchers/any-args?) mockfn.macros/fall-through
```

I also
 - implemented `any-args?` (appropriate name?) to match irregardless of # of args provided
 - deprecated `unmocked` and replaced it with `fall-through`

philomates committed:
fix cljs tests/impl of unbound replacment in message

dchelimsky commented on doc/documentation.md:
```diff
@@ -59,14 +59,14 @@ If you would like to call the value instead of returning it, use `mockfn.macros/
+#### `fall-through`
 
 When mocking it is sometimes useful to set some mocks to point to their
-original implementation. This can be done by using `mockfn.macros/unmocked`:
+original implementation. This can be done by using `mockfn.macros/fall-through`:
```
How about:

```clojure
When mocking a function, it is sometimes useful to allow calls with specific arguments to fall through
to their original implementation:
```

philomates committed:
reword

dchelimsky reviewed (approved):
Thanks @philomates !

philomates committed:
changelog + version bump
